### PR TITLE
mds: don't stall the asok thread for flush commands

### PR DESCRIPTION
**Pull request by leonid-s-usov**
Also, relax some timing requirements in the quiescer

Fixes: https://tracker.ceph.com/issues/65803

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by leonid-s-usov** [comment]
This is used as the suite branch for https://pulpito.ceph.com/leonidus-2024-05-05_21:28:36-fs-wip-pdonnell-testing-20240503.163550-debug-distro-default-smithi/

**Review by vshankar — COMMENTED** [review]
Looks fine!

Conditional approve post successful test run.

**Review by batrick — CHANGES_REQUESTED** [review]
https://tracker.ceph.com/issues/65803:#note-2

I think the simple fix is to make `flush journal` asynchronous. The `quiesce` command already is so please just do some similar logic there.

**Comment by leonid-s-usov** [comment]
> https://tracker.ceph.com/issues/65803:#note-2
> 
> I think the simple fix is to make `flush journal` asynchronous. The `quiesce` command already is so please just do some similar logic there.

I'm not sure it's that simple. We can't just change a synchronous command to asynchronous, can we? We'll have to add a switch to the command and change the invocation along with verifying that it'll know how to deal with the new approach.

In any case, it would probably have to be a separate ticket with its own PR...

**Comment by leonid-s-usov** [comment]
> > https://tracker.ceph.com/issues/65803:#note-2
> > I think the simple fix is to make `flush journal` asynchronous. The `quiesce` command already is so please just do some similar logic there.
> 
> I'm not sure it's that simple. We can't just change a synchronous command to asynchronous, can we? We'll have to add a switch to the command and change the invocation along with verifying that it'll know how to deal with the new approach.
> 
> In any case, it would probably have to be a separate ticket with its own PR...

In a talk with @batrick he explained that it was about not holding up the asok thread, which I followed here

**Review by batrick — APPROVED** [review]

**Comment by leonid-s-usov** [comment]
jenkins test make check

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -2949,9 +2957,17 @@ void MDSRankDispatcher::handle_asok_command(
...
+
+    std::lock_guard l(mds_lock);
+    mdcache->flush_dentry(path, new AsyncResponse(f, std::move(on_finish)));
+    return;
   } else if (command == "flush journal") {
-    command_flush_journal(f);
+    auto respond = new AsyncResponse(f, std::move(on_finish));
+    C_Flush_Journal* flush_journal = new C_Flush_Journal(mdcache, mdlog, this, &respond->css, respond);
+
+    std::lock_guard locker(mds_lock);
```
Comment (lxbsz) [inline src/mds/MDSRank.cc:2968]: We are always suggested to use the `scoped_lock` instead. For example in https://github.com/ceph/ceph/pull/36221#discussion_r463856968.

Comment (leonid-s-usov) [reply src/mds/MDSRank.cc:2968]: My research didn't explain why, would you please elaborate?
For a single lock case `lock_guard` is still good: https://stackoverflow.com/a/60172828/5171225

**Comment by leonid-s-usov** [comment]
jenkins test make check

File: src/mds/MDSRank.cc (hunk 1 above)
Comment (batrick) [reply src/mds/MDSRank.cc:2968]: Yes, it's my fault proliferating that. I had recently seen that discussion comment too. I would consider `lock_guard` to be a good default choice going forward.

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/mds/MDSRank.cc (hunk 1 above)
Comment (lxbsz) [reply src/mds/MDSRank.cc:2968]: I couldn't remember the another discussing PR about this. Since it was @batrick's idea then I am okay with it.

**Review by lxbsz — APPROVED** [review]

**Review by vshankar — CHANGES_REQUESTED** [review]
Nicely done.

**Review by vshankar — APPROVED** [review]
Approved!

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/65867](https://tracker.ceph.com/issues/65867).

**Comment by leonid-s-usov** [comment]
@batrick please approve for merge. See https://github.com/ceph/ceph/pull/57332#issuecomment-2111021712

**Pull request merged**